src/fabric/src/fabric_rpc.erl

```diff
@@ -298,6 +314,104 @@ with_db(DbName, Options, {M,F,A}) ->
         rexi:reply(Error)
     end.
 
+
+read_repair_filter(DbName, Docs, NodeRevs, Options) ->
+    set_io_priority(DbName, Options),
+    case get_or_create_db(DbName, Options) of
+        {ok, Db} ->
+            try
+                read_repair_filter(Db, Docs, NodeRevs)
+            after
+                couch_db:close(Db)
+            end;
+        Error ->
+            rexi:reply(Error)
+    end.
+
+
+% A read repair operation may have been triggered by a node
+% that was out of sync with the local node. Thus, any time
+% we receive a read repair request we need to check if we
+% may have recently purged any of the given revisions and
+% ignore them if so.
+%
+% This is accomplished by looking at the purge infos that we
+% have locally that have not been replicated to the remote
+% node. The logic here is that we may have received the purge
+% request before the remote shard copy. So to check that we
+% need to look at the purge infos that we have locally but
+% have not yet sent to the remote copy.
+%
+% NodeRevs is a list of the {node(), [rev()]} tuples passed
+% as the read_repair option to update_docs.
+read_repair_filter(Db, Docs, NodeRevs) ->
+    [#doc{id = DocId} | _] = Docs,
+    Nodes = lists:usort([Node || {Node, _} <- NodeRevs, Node /= node()]),
+    NodeSeqs = get_node_seqs(Db, Nodes),
```

Comment:
I would use `gb_sets` here since the number of elements returned by `couch_db:fold_local_docs` inside `get_node_seq` can be big. It seems we need to get `NodeSeq` for given Node quite a few times here. We do it in the loop when calculate `RecentNodeRevs` and in another loop when we calculate `AllowableRevs`.

Reply:
Feel free to ignore if you are sure that the number of elements in `NodeSeqs` list is always low (less than 1000).

Reply:
Nodes from line 349 is capped at 2 (or more specifically N-1) so performance should be acceptable for any reasonable N.

Reply:
Also, the reason for that is that this is a read repair is only ever called from fabric_doc_open and fabric_doc_open_revs and we're just sending the revisions that we get from each node in the response. And worst case would be every node in the cluster if we ever improved that algorithm to do parallel opens.